Изможеров Илья. V2.0

TrueCraftWorld opened the pull request:
Version 2.0 - Main difference - only 1 Qtimer for drop control instead of 1 for each tile.

TrueCraftWorld committed:
1st realisation. No comments, no speedup at hover

TrueCraftWorld committed:
Adding: speedup on hover, colors of Box

TrueCraftWorld committed:
testin performance fixes

TrueCraftWorld committed:
alternate approach to speed

TrueCraftWorld committed:
testin performance fixes

TrueCraftWorld committed:
minor Fixes

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+
+/// Your code here...
+
+    const int tile_size = 30;
```
Можно уж `constexpr` тогда.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+
+    static bool isLost = false;
+
+    int butNum = 0;
```
Это явно лишнее.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+    auto bigDropTimer = new QTimer(this);
+    QMap<QString, int> dropStepMap;
+
+    this->setWindowTitle(QString("SQUARE FALLDOWN"));
```
`this->` как правило лишнее, за редким исключением.

Toxin65 replied:
Не учтено, до сих пор.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+    QObject::connect(deployTimer,
+                     &QTimer::timeout,
+                     this,
+                     [=]() mutable {
```
`[=]` Так делать очень плохо. Захватывать нужно необходимый минимум. Захват всего и вся, тем более по значению, очень нехорошо.
Ещё и `mutable`.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+        QPushButton* newButton = new QPushButton(this);
+        newButton->setAccessibleName(QString("But%1").arg(butNum));
+        butNum++;
+        int speed = QRandomGenerator::global()->bounded(lowestSpeed,highestSpeed);
```
Я, наверное, не говорил. `const` лучше писать всегда, кроме тех случаев когда это действительно нельзя.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+        int speed = QRandomGenerator::global()->bounded(lowestSpeed,highestSpeed);
+        QString test(newButton->accessibleName());
+
+        dropStepMap.insert(test, speed);
```
Оверкомлексити... Слишком сложно. Не нужно.

TrueCraftWorld replied:
Это про использование этого контейнера? Показалось нормальным хранить пары <имя, скорость> и map подходит под хранение пар. 
Будет ли нормально тогда спрятать скорость в какой-то параметр дефолтной QPudhButton, например, в имя?

Toxin65 replied:
Ниже ответил "Просто захватывайте скорость в лямбду. "

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+                                QRandomGenerator::global()->bounded(255),
+                                QRandomGenerator::global()->bounded(100, 255))); //to make them colorfull but with no chance of being same as BG
+        newButton->setAutoFillBackground(true);
+        newButton->setPalette(palette1);
```
Это не требовалось по ТЗ. Лишние строки. Нууу, если хотите, пусть будет.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+        newButton->setGeometry(QRect(QRandomGenerator::global()->bounded(this->width()-tile_size),
+        highBorder,
+        tile_size,
+        tile_size));
```
Переносы по кодстайлу выглядят ужасно. Ужасно нечитаемо.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+        tile_size,
+        tile_size));
+
+        QObject::connect(bigDropTimer, &QTimer::timeout,  newButton, [=] {
```
Использование единого таймера, безусловно интересный подход.
По захватам то же замечание.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+
+        QObject::connect(bigDropTimer, &QTimer::timeout,  newButton, [=] {
+                int step = (newButton->underMouse()) ? 2*dropStepMap[newButton->accessibleName()] :
+                                                        dropStepMap[newButton->accessibleName()];
```
Дорахо бохато. Не нужно. Просто захватывайте скорость в лямбду. Ведь и так на каждую кнопку своя лямбда. Пусть оно живёт не в `dropStepMap`, и доступ значительно быстрее.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+                int step = (newButton->underMouse()) ? 2*dropStepMap[newButton->accessibleName()] :
+                                                        dropStepMap[newButton->accessibleName()];
+                newButton->move(newButton->x(),newButton->y()+step);
+                if (newButton->y() >= this->height() - tile_size) {
```
Если вы хотите заморочится, то сначала "короткую ветку". Условие инвертировать и `return`. И ниже то же.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+                    newButton->deleteLater();
+                }
+        } );
+        QObject::connect(newButton, &QPushButton::clicked, newButton, [=] { //makes it easier to play
```
Ну вот тут захваты просто ужас, тут только кнопка нужна.
И ещё. Тут не нужна лямбда -- напрямую сигнал в `deleteLater` направлять.

TrueCraftWorld replied:
Разве должна кнопка удалятся по клику, когда уже проиграли? Лямбда только из-за этой проверки здесь.

Toxin65 replied:
Этого в ТЗ не сказано. Значит без разницы.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,95 @@
+    } );
+
+    QObject::connect(deployTimer, &QTimer::timeout,  this, [=] {
+        deployTimer->setInterval(QRandomGenerator::global()->bounded(minSpawnInterval, maxSpawnInterval));
```
А почему это не сделать в лямбде генерации?

TrueCraftWorld replied:
Перенёс в генерацию

Toxin65 reviewed (changes requested):
В целом нормально. Но есть серьёзные замечания.

TrueCraftWorld committed:
прописал захваты.
 удалил лишнее.
 изменил способ хранения скорости.
для констант настроек добавил constexpr

TrueCraftWorld committed:
Merge branch 'test2' into test

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,81 @@
+    this->setWindowTitle(QString("SQUARE FALLDOWN"));
+    auto deployTimer = new QTimer();
+    deployTimer->setInterval(QRandomGenerator::global()->bounded(minSpawnInterval, maxSpawnInterval));
+    auto bigDropTimer = new QTimer();
```
Ой ой ой. А почему у таймеров владельцы пропали??? Было правильно.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,81 @@
+    QObject::connect(deployTimer,
+                     &QTimer::timeout,
+                     this,
+                     [this, highBorder, tile_size, bigDropTimer, deployTimer]() {
```
`highBorder, tile_size` лишние. Подумайте почему.
`m_timer` есть вместо для `deployTimer`, тогда его не надо захватывать отдельно.

TrueCraftWorld replied:
Потому что они теперь констэкпр и видны "глобально". Не продумал, когда им менял объявление.

TrueCraftWorld replied:
Про m_timer не понял. Имеется в виду вызываемый через startTimer? ,чтобы его вместе с this захватить? А как тогда ловить его сигналы - дока пишет что у него обработчик переопределяется наследником, а нам же так нельзя по ТЗ.

Toxin65 replied:
Верно.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,81 @@
+        tile_size,
+        tile_size));
+
+        QObject::connect(bigDropTimer, &QTimer::timeout,  newButton, [this, newButton, tile_size] {
```
`tile_size` лишнее.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,81 @@
+
+        QObject::connect(bigDropTimer, &QTimer::timeout,  newButton, [this, newButton, tile_size] {
+
+            int step = newButton->accessibleName().toInt();
```
Не надо так делать, захватывайте в лямбду.

TrueCraftWorld replied:
Т.е. для каждой кнопки своя личная лямбда с копией того значения, что захватим? 
Не осознавал, что это так работает.

Toxin65 replied:
Она по факту для каждого своя, да. Для каждого объекта свой экземпляр. Поэтому не надо этих излишеств.

Toxin65 reviewed (changes requested):
Появились новые замечания :(

TrueCraftWorld committed:
lambda and constexpr fixes

TrueCraftWorld committed:
m_timer usage

Toxin65 commented on mainwindow.cpp:
```diff
@@ -24,29 +24,29 @@ MainWindow::MainWindow(QWidget *parent)
     bigDropTimer->setInterval(100);
 
-    QObject::connect(deployTimer,
+    QObject::connect(this->m_timer,
```
Да штош такое, питон вас покусал... Просто `m_timer`. `this` для этого не нужен.

TrueCraftWorld replied:
Да, ускользает от взора...

Оба случая исправил

TrueCraftWorld committed:
this->m_timer fix

TrueCraftWorld committed:
another this-> fix

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,81 @@
+        QPushButton* newButton = new QPushButton(this);
+        const int speed = QRandomGenerator::global()->bounded(lowestSpeed,highestSpeed);
+        newButton->setGeometry(QRect(QRandomGenerator::global()->bounded(width()-tile_size),
+                                        highBorder,
```
Но ведь не соответствует ТЗ же!
`- Объекты появляются в произвольном месте по всей ширине окна и в области 100 пикселей от верха окна;`
Ладно, ТЗ в этом моменте, мутноватое.

Toxin65 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,81 @@
+        QObject::connect(bigDropTimer, &QTimer::timeout,  newButton, [this, newButton, speed] {
+
+            int step = speed;
+            if (newButton->underMouse()) step*=2;
```
Можно тернарником и сразу в конст.

TrueCraftWorld committed:
spawn fixes
